HouseRules_Essentials/Rules/MonsterDeckOverriddenRule.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            if (!_isActivated)
+            {
+                return true;
+            }
+
+            List<MonsterDeck.MonsterDeckEntry> standardDeck;
+            List<MonsterDeck.MonsterDeckEntry> spikeDeck;
+            MonsterDeck.MonsterDeckEntry keyHolder;
+            if (floorIndex == 1)
+            {
+                standardDeck = CreateSubDeck(_globalAdjustments.EntranceDeckFloor1);
+                spikeDeck = CreateSubDeck(_globalAdjustments.ExitDeckFloor1);
+                keyHolder = new MonsterDeck.MonsterDeckEntry() { BoardPieceId = _globalAdjustments.KeyHolderFloor1, enemyWeight = EnemyWeight.Light, isRedrawEnabled = false };
+            }
+            else
+            {
+                standardDeck = CreateSubDeck(_globalAdjustments.EntranceDeckFloor2);
+                spikeDeck = CreateSubDeck(_globalAdjustments.ExitDeckFloor2);
+                keyHolder = new MonsterDeck.MonsterDeckEntry() { BoardPieceId = _globalAdjustments.KeyHolderFloor2, enemyWeight = EnemyWeight.Light, isRedrawEnabled = false };
+            }
+
+            List<MonsterDeck.MonsterDeckEntry> bossDeck = CreateSubDeck(_globalAdjustments.BossDeck);
+
+            MonsterDeck monsterDeck = new MonsterDeck
+            {
+                monsterDeckStandard = standardDeck,
+                monsterDeckSpike = spikeDeck,
+                monsterDeckBoss = bossDeck,
+                Keyholder = keyHolder,
+            };
+            monsterDeck.ShuffleSubDecks(rng);
+            __result = monsterDeck;
+            return false; // We returned an user-adjusted config.
+        }
+
+        private static bool AIDirectorDeckConstructor_AddPreFillUnitsToDeck_Prefix(List<MonsterDeck.MonsterDeckEntry> monsterDeckList, ISpawnCategoryProvider spawnCategoryProvider)
+        {
+            if (!_isActivated)
+            {
+                return true;
+            }
+
+            return false; // We did all that we needed to do, and that was nothing.
+        }
+
+        private static bool AIDirectorController2_SpawnBossAndMinions_Prefix(ref AIDirectorController2 __instance, ref AIDirectorContext context, IRnd rng, ref TransientBoardState boardState)
+        {
+            if (!_isActivated)
+            {
+                return true;
+            }
+
+            SpawnZoneTag tag = SpawnZoneTag.Zone2;
+            SpawnZone spawnZone = null;
+            SpawnZone spawnZone2 = null;
+            List<SpawnZone> list;
+            list = context.spawnZoneModel.GetZonesWithTag(tag);
+            for (int i = list.Count - 1; i >= 0; i--)
+            {
+                SpawnZone spawnZone3 = list[i];
+                if (spawnZone3.GetAllFreeTiles(ref boardState).Count == 0)
+                {
+                    EssentialsMod.Logger.Msg(string.Format("Spawn zone doesn't contain any free tiles, removing it! {0}", spawnZone3.TileRect), string.Empty);
```

Comment:
IIUC, I think `string.Empty` can be removed from the end.

Or consider using string interpolation for nicer strings:
```cs
EssentialsMod.Logger.Msg($"Spawn zone doesn't contain any free tiles, removing it! {spawnZone3.TileRect}");
```